File: focus-log/src/lib/conformance.test.ts
/**
 * Cross-core conformance: the web core must agree with the Wear OS (Kotlin) core
 * on every rule the server-free last-write-wins protocol depends on. Both suites
 * load the *same* vectors in /conformance; the Kotlin twin is
 * focus-log-wear/core/.../ConformanceTest.kt.
 *
 * If this drifts from the Kotlin core, two devices can disagree on which edit
 * won or how long a session was — the exact class of bug the vectors exist to
 * catch. Keep the vectors as the shared source of truth; change both cores
 * together.
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { elapsedSeconds, type Segment, type TimerState } from "./timer/engine";
import { compareVersions, pickWinner, type Versioned } from "./sync/merge";
import { decodeSegments, encodeSegments } from "./sheets/schema";
import {
  closedActive,
  finalizedSession,
  reconcileActive,
  runningActive,
  timerFromActive,
} from "./timer/lifecycle";
import type { ActiveTimer } from "./sheets/schema";

function loadVectors<T>(file: string): T {
  const path = fileURLToPath(new URL(`../../../conformance/${file}`, import.meta.url));
  return JSON.parse(readFileSync(path, "utf8")) as T;
}

/** A [startMs, endMs|null] pair as stored in the vectors. */
type SegPair = [number, number | null];
const toSegments = (pairs: SegPair[]): Segment[] => pairs.map(([start, end]) => ({ start, end }));

describe("conformance: elapsedSeconds", () => {
  const { cases } = loadVectors<{
    cases: { name: string; segments: SegPair[]; now: number; expectedSeconds: number }[];
  }>("elapsed.json");

  it.each(cases)("$name", ({ segments, now, expectedSeconds }) => {
    const state = { goalId: "g", segments: toSegments(segments), startedAt: 0, note: "" };
    expect(elapsedSeconds(state, now)).toBe(expectedSeconds);
  });
});

describe("conformance: last-write-wins", () => {
  type Rec = Versioned & { id: string };
  const { cases } = loadVectors<{
    cases: { name: string; a: Rec; b: Rec; winner: "a" | "b" }[];
  }>("lww.json");

  it.each(cases)("$name", ({ a, b, winner }) => {
    const ra: Rec = { ...a, deleted: false };
    const rb: Rec = { ...b, deleted: false };
    const cmp = compareVersions(ra, rb);
    expect(pickWinner(ra, rb).id).toBe(winner === "a" ? "A" : "B");

    if (cmp !== 0) {
      // A real winner is order-independent: the same record wins either way.
      expect(pickWinner(rb, ra).id).toBe(winner === "a" ? "A" : "B");
      expect(winner === "b" ? cmp < 0 : cmp > 0).toBe(true);
    } else {
      // An exact tie is expressed as winner "a", and pickWinner deterministically
      // returns its first argument (so the swap returns "B").
      expect(winner).toBe("a");
      expect(pickWinner(rb, ra).id).toBe("B");
    }
  });
});

describe("conformance: segments codec", () => {
  const { cases, invalidEncoded } = loadVectors<{
    cases: { name: string; segments: SegPair[]; encoded: string }[];
    invalidEncoded: string[];
  }>("segments-codec.json");

  it.each(cases)("$name", ({ segments, encoded }) => {
    expect(encodeSegments(toSegments(segments))).toBe(encoded);
    expect(decodeSegments(encoded)).toEqual(toSegments(segments));
  });

  it.each(invalidEncoded)("rejects %j", (bad) => {
    expect(() => decodeSegments(bad)).toThrow();
  });
});

describe("conformance: active-timer lifecycle mapping", () => {
  interface MappingCase {
    name: string;
    input: {
      logId: string;
      goalId: string;
      segments: SegPair[];
      note: string;
      tz: string;
      deviceId: string;
      now: number;
    };
    finalizedSession: Record<string, unknown>;
    activeEncodedSegments: string;
    activeUpdatedAt: string;
  }
  const { cases } = loadVectors<{ cases: MappingCase[] }>("active-mapping.json");

  it.each(cases)(
    "$name",
    ({ input, finalizedSession: expected, activeEncodedSegments, activeUpdatedAt }) => {
      const segments = toSegments(input.segments);
      const state: TimerState = {
        goalId: input.goalId,
        segments,
        startedAt: segments[0]?.start ?? 0,
        note: input.note,
      };
      const ctx = { logId: input.logId, deviceId: input.deviceId, now: input.now };

      expect(finalizedSession(state, { ...ctx, tz: input.tz })).toEqual(expected);

      const active = runningActive(state, ctx);
      expect(encodeSegments(active.segments)).toBe(activeEncodedSegments);
      expect(active.updated_at).toBe(activeUpdatedAt);
      expect(active.deleted).toBe(false);
      expect(active.log_id).toBe(input.logId);
      expect(closedActive(state, ctx).deleted).toBe(true);

      // Round-trip: a device adopting the published row rebuilds the same timer.
      expect(timerFromActive(active)).toEqual(state);
    },
  );
});

describe("conformance: reconcileActive", () => {
  interface ActiveJson {
    log_id: string;
    goal_id: string;
    segments: SegPair[];
    note: string;
    updated_at: string;
    deleted: boolean;
    device_id: string;
  }
  const toActive = (a: ActiveJson | null): ActiveTimer | undefined =>
    a ? { ...a, segments: toSegments(a.segments) } : undefined;

  interface ReconcileCase {
    name: string;
    local: ActiveJson | null;
    remote: ActiveJson[];
    expected: { local: ActiveJson | null; closeLogId: string | null; changed: boolean };
  }
  const { cases } = loadVectors<{ cases: ReconcileCase[] }>("reconcile.json");

  it.each(cases)("$name", ({ local, remote, expected }) => {
    const result = reconcileActive(toActive(local), remote.map(toActive) as ActiveTimer[]);
    expect(result.local).toEqual(toActive(expected.local));
    expect(result.closeLogId ?? null).toBe(expected.closeLogId);
    expect(result.changed).toBe(expected.changed);
  });
});

File: focus-log-wear/core/src/test/kotlin/com/focuslog/core/ConformanceTest.kt
package com.focuslog.core

import com.focuslog.core.model.Versioned
import com.focuslog.core.sheets.decodeSegments
import com.focuslog.core.sheets.encodeSegments
import com.focuslog.core.sync.compareVersions
import com.focuslog.core.sync.pickWinner
import com.focuslog.core.timer.Segment
import com.focuslog.core.timer.TimerEngine
import com.focuslog.core.timer.TimerState
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File

/**
 * Cross-core conformance: the Wear OS core must agree with the web (TypeScript)
 * core on every rule the server-free last-write-wins protocol depends on. Both
 * suites load the *same* vectors in /conformance; the TS twin is
 * focus-log/src/lib/conformance.test.ts.
 *
 * If this drifts from the web core, two devices can disagree on which edit won or
 * how long a session was. Keep the vectors the shared source of truth and change
 * both cores together. The /conformance path is passed in by Gradle as the
 * `conformanceDir` system property (see core/build.gradle.kts).
 */
class ConformanceTest {
    private val dir = File(
        System.getProperty("conformanceDir") ?: error("conformanceDir system property not set"),
    )

    private fun load(file: String): Map<String, Any?> {
        @Suppress("UNCHECKED_CAST")
        return MiniJson.parse(File(dir, file).readText()) as Map<String, Any?>
    }

    private fun segmentsOf(pairs: Any?): List<Segment> {
        @Suppress("UNCHECKED_CAST")
        return (pairs as List<Any?>).map {
            @Suppress("UNCHECKED_CAST")
            val pair = it as List<Any?>
            Segment((pair[0] as Double).toLong(), (pair[1] as Double?)?.toLong())
        }
    }

    @TestFactory
    fun elapsed(): List<DynamicTest> {
        @Suppress("UNCHECKED_CAST")
        val cases = load("elapsed.json")["cases"] as List<Map<String, Any?>>
        return cases.map { c ->
            DynamicTest.dynamicTest(c["name"] as String) {
                val state = TimerState("g", segmentsOf(c["segments"]), startedAt = 0L, note = "")
                val now = (c["now"] as Double).toLong()
                assertEquals((c["expectedSeconds"] as Double).toLong(), TimerEngine.elapsedSeconds(state, now))
            }
        }
    }

    private data class Rec(
        val id: String,
        override val updatedAt: String,
        override val deviceId: String,
        override val deleted: Boolean = false,
    ) : Versioned

    private fun recOf(m: Map<String, Any?>): Rec =
        Rec(m["id"] as String, m["updated_at"] as String, m["device_id"] as String)

    @TestFactory
    fun lastWriteWins(): List<DynamicTest> {
        @Suppress("UNCHECKED_CAST")
        val cases = load("lww.json")["cases"] as List<Map<String, Any?>>
        return cases.map { c ->
            DynamicTest.dynamicTest(c["name"] as String) {
                @Suppress("UNCHECKED_CAST")
                val a = recOf(c["a"] as Map<String, Any?>)
                @Suppress("UNCHECKED_CAST")
                val b = recOf(c["b"] as Map<String, Any?>)
                val winner = c["winner"] as String
                val expectedId = if (winner == "a") "A" else "B"
                assertEquals(expectedId, pickWinner(a, b).id)

                if (compareVersions(a, b) != 0) {
                    // A real winner is order-independent.
                    assertEquals(expectedId, pickWinner(b, a).id)
                } else {
                    // An exact tie is expressed as "a"; pickWinner returns its first arg.
                    assertEquals("a", winner)
                    assertEquals("B", pickWinner(b, a).id)
                }
            }
        }
    }

    @TestFactory
    fun segmentsCodec(): List<DynamicTest> {
        val data = load("segments-codec.json")
        @Suppress("UNCHECKED_CAST")
        val cases = data["cases"] as List<Map<String, Any?>>
        @Suppress("UNCHECKED_CAST")
        val invalid = data["invalidEncoded"] as List<String>

        val roundTrips = cases.map { c ->
            DynamicTest.dynamicTest(c["name"] as String) {
                val segments = segmentsOf(c["segments"])
                val encoded = c["encoded"] as String
                assertEquals(encoded, encodeSegments(segments))
                assertEquals(segments, decodeSegments(encoded))
            }
        }
        val rejections = invalid.map { bad ->
            DynamicTest.dynamicTest("rejects \"$bad\"") {
                assertThrows(Exception::class.java) { decodeSegments(bad) }
            }
        }
        return roundTrips + rejections
    }
}
